### ISSUE TITLE
Discussion: Should reindex keep the index name if new index has none

### ISSUE THREAD

**Issue by hsperr** [pandas-dev/pandas#9885]
As mentioned in #6552 there is an open question about whether reindex should keep the index name if the new index has no name.

Current Behavior:
- reindex keeps index name if the argument is a list or an ndarray
- replaces the index name if the argument is an index, even if the new index does not have a new name.

Related Issues:
#9854

**Comment by TomAugspurger** [comment]
There isn't an obvious answer to me for this one.

Also keep in mind that we have `reindex_like`. That one isn't clear either, but they should probably have the same behavior? Maybe?

**Comment by shoyer** [comment]
I think `reindex` and `reindex_like` should definitely have the same behavior.

We do technically associate the name of an index with the index itself, but from a practical perspective the name of an index is also strongly associated with the dataframe that contains it (somewhat similarly to the case for `Series.name` and `DataFrame.columns`). In my opinion, it is a little surprising (even if it happens rarely) for reindexing to rename existing indexes -- it makes it harder to understand how the data was transformed. So my vote is for reindexing to *never* change index names.

**Comment by attack68** [comment]
I'm developing an application which uses this feature. I came across page researching a bug in my application. I have used re_index to readjust on the fly to a function local Index variable which I did not bother to assign a name to. Reindex therefore dropped my index name and it broke down the line.

I would suggest that the index name is preserved if:
> the reindex argument is a list, ndarray or Index which does not have a name,

and the index name is renamed if:
> the reindex argument is an Index which has a specific name.

Perhaps an optional 'name_transform=None' argument could be added with the default behaviour as above and further 'drop' or 'keep' options to explicitly define the behaviour.
